helper/keysutil/lock_manager.go

```diff
@@ -103,26 +103,50 @@ func (lm *LockManager) RestorePolicy(ctx context.Context, storage logical.Storag
 	lock.Lock()
 	defer lock.Unlock()
 
-	// If the policy is in cache, error out. Anywhere that would put it in the
-	// cache will also be protected by the mutex above, so we don't need to
-	// re-check the cache later.
-	_, ok := lm.cache.Load(name)
-	if ok {
+	// If the policy is in cache and 'force' is not specified, error out. Anywhere
+	// that would put it in the cache will also be protected by the mutex above,
+	// so we don't need to re-check the cache later.
+	pRaw, ok := lm.cache.Load(name)
+	if ok && !force {
 		return fmt.Errorf(fmt.Sprintf("key %q already exists", name))
 	}
 
-	// If the policy exists in storage, error out
-	p, err := lm.getPolicyFromStorage(ctx, storage, name)
-	if err != nil {
-		return err
+	// Conditionally look up the policy from storage, depending on the use of
+	// 'force' and if the policy was found in cache.
+	//
+	// - If was not found in cache and we are not using 'force', look for it in
+	// storage. If found, error out.
+	//
+	// - If it was found in cache and we are using 'force', pRaw will not be nil
+	// and we do not look the policy up from storage
+	//
+	// - If it was found in cache and we are not using 'force', we should have
+	// returned above wih error
+	var p *Policy
+	if pRaw == nil {
+		p, err = lm.getPolicyFromStorage(ctx, storage, name)
+		if err != nil {
+			return err
+		}
+		if p != nil && !force {
+			return fmt.Errorf(fmt.Sprintf("key %q already exists", name))
+		}
+	}
+
+	// If both pRaw and p above are nil and 'force' is specified, we don't need to
+	// grab policy locks as we have ensured it doesn't already exist, so there
+	// will be no races as nothing else has this pointer. If 'force' was not used,
+	// an error would have been returned by now if the policy already existed
+	if pRaw != nil {
+		p = pRaw.(*Policy)
 	}
 	if p != nil {
-		return fmt.Errorf(fmt.Sprintf("key %q already exists", name))
+		p.l.Lock()
+		defer func() {
```

Comment:
Why is an anonymous function needed for this `defer`?

Reply:
Probably bad copy-paste 😦